lib/change_stream.js

```diff
@@ -0,0 +1,283 @@
+ * @class ChangeStream
+ * @since 3.0.0
+ * @param {(Db|Collection)} changeDomain The database or collection against which to create the change stream
+ * @param {Array} pipeline An array of {@link https://docs.mongodb.com/manual/reference/operator/aggregation-pipeline/|aggregation pipeline stages} through which to pass change stream documents
+ * @param {object} [options=null] Optional settings
+ * @param {string} [options.fullDocument=none] Allowed values: ‘none’, ‘lookup’. When set to ‘lookup’, the change stream will include both a delta describing the changes to the document, as well as a copy of the entire document that was changed from some time after the change occurred.
+ * @param {number} [options.maxAwaitTimeMS] The maximum amount of time for the server to wait on new documents to satisfy a change stream query
+ * @param {object} [options.resumeAfter=null] Specifies the logical starting point for the new change stream. This should be the _id field from a previously returned change stream document.
+ * @param {number} [options.batchSize=null] The number of documents to return per batch. See {@link https://docs.mongodb.com/manual/reference/command/aggregate|aggregation documentation}.
+ * @param {object} [options.collation=null] Specify collation settings for operation. See {@link https://docs.mongodb.com/manual/reference/command/aggregate|aggregation documentation}.
+ * @param {ReadPreference} [options.readPreference=null] The read preference. Defaults to the read preference of the database or collection. See {@link https://docs.mongodb.com/manual/reference/read-preference|read preference documentation}.
+ * @fires ChangeStream#close
+ * @fires ChangeStream#data
+ * @fires ChangeStream#end
+ * @fires ChangeStream#error
+ * @return {ChangeStream} a ChangeStream instance.
+ */
+var ChangeStream = function(changeDomain, pipeline, options) {
+  var Collection = require('./collection');
+  var Db = require('./db');
+
+  var self = this;
+  self.changeDomain = changeDomain;
+  self.pipeline = pipeline || [];
+  self.options = options || {};
+
+  // Extract namespace and serverConfig from changeDomain
+  self.namespace = {};
+  var changeDb;
+  if (changeDomain instanceof Collection) {
+    self.namespace.collection = changeDomain.collectionName;
+    changeDb = changeDomain.s.db;
+  } else if (changeDomain instanceof Db) {
+    changeDb = changeDomain;
+  } else {
+    throw new Error('changeDomain must be either a database or collection.');
+  }
+  self.namespace.database = changeDb.databaseName;
+  self.serverConfig = changeDb.serverConfig;
+
+  // Determine correct read preference
+  self.options.readPreference = self.options.readPreference || changeDomain.readPreference || changeDomain.s.readPreference;
+
+  // Create contained Change Stream cursor
+  self.cursor = createChangeStreamCursor(self);
+
+  self.on('newListener', function(eventName) {
```

Comment:
You are correct that there are two paths related to `data` listeners. The path on line 54 facilitates `data` listeners added after the contained cursor's creation, while the path on line 87 relates to `data` listeners on the ChangeStream object at the time of contained cursor's creation.

The reason I've done this is because placing a `data` listener on ChangeStream's contained cursor switches that cursor to flowing mode (see [Node Stream docs](https://nodejs.org/api/stream.html#stream_two_modes)). In flowing mode, new data is immediately broadcast to any listeners. If no-one is listening to the event, then the data will be lost.

Because of this nuance, to prevent data loss we only want to place a `data` listener on the contained cursor if there is a `data` listener on the ChangeStream object.

If you can think of a cleaner way to do this, I'd be happy to hear it!